src/main/java/bio/terra/workspace/service/resource/controlled/ControlledResourceService.java

```diff
@@ -31,17 +36,59 @@
   private final WorkspaceService workspaceService;
   private final ResourceDao resourceDao;
   private final StageService stageService;
+  private final SamService samService;
 
   @Autowired
   public ControlledResourceService(
       JobService jobService,
       WorkspaceService workspaceService,
       ResourceDao resourceDao,
-      StageService stageService) {
+      StageService stageService,
+      SamService samService) {
     this.jobService = jobService;
     this.workspaceService = workspaceService;
     this.resourceDao = resourceDao;
     this.stageService = stageService;
+    this.samService = samService;
+  }
+
+  /**
+   * Convenience function that checks existence of a controlled resource within a workspace,
+   * followed by an authorization check against that resource.
+   *
+   * <p>Throws ResourceNotFound from getResource if the workspace does not exist in the specified
+   * workspace, regardless of the user's permission.
+   *
+   * <p>Throws InvalidControlledResourceException if the given resource is not controlled.
+   *
+   * <p>Throws SamUnauthorizedException if the user is not permitted to perform the specified action
+   * on the resource in question.
+   *
+   * <p>Returns the controlled resource object if it exists and the user is permitted to perform the
+   * specified action.
+   *
+   * @param userReq the user's authenticated request
+   * @param workspaceId if of the workspace this resource exists in
+   * @param resourceId id of the resource in question
+   * @param action the action to authorize against the resource
+   * @return the resource, if it exists and the user is permitted to perform the specified action.
+   */
+  @Traced
+  public ControlledResource validateControlledResourceAndAction(
+      AuthenticatedUserRequest userReq, UUID workspaceId, UUID resourceId, String action) {
+    WsmResource resource = resourceDao.getResource(workspaceId, resourceId);
+    if (resource.getStewardshipType() != StewardshipType.CONTROLLED) {
+      throw new InvalidControlledResourceException(
```

Comment:
We know the name of the resource, too, which might be more useful to humans?
Should this always be combined with a "do you have any access to the workspace?" check?

Reply:
Added resource ID to the error here. If the resource is not found in the given workspace, `resourceDao.getResource` will throw a `ResourceNotFoundException`

The name is probably more useful, but I think names for references aren't validated atm. Since they're user-provided, we probably shouldn't echo them back

Added TODO for PF-640. I think we probably want that check for workspace read access, but I want to be consistent about fixing private users / workspace access